File: src/receipts.ts
// Receipts for the client you already use. Every mainstream client accepts an
// injected `fetch`; this one reads the `x-conifer-*` headers on the way past
// and hands the response back untouched. The body is never read here.
//
//   const receipts = new ReceiptCollector();
//   const openai = new OpenAI({ baseURL, apiKey, fetch: receipts.fetch });
//   await openai.chat.completions.create({ ... });
//   receipts.total.costNanoUsd;

import { nanoUsdToUsdString, readReceipt, type Receipt } from "./receipt.ts";
import type { FetchLike } from "./transport.ts";

export interface ObservedReceipt extends Receipt {
  /** The URL that was called. */
  url: string;
  /** When the response arrived. */
  at: Date;
}

export interface ReceiptTotal {
  /** Turns that disclosed a cost. */
  turns: number;
  /** Summed settled cost, in integer nanodollars. */
  costNanoUsd: number;
  /** The same as an exact decimal USD string. */
  costUsd: string;
  /** Summed counterfactual over the turns that disclosed one; a subset, not a savings figure. */
  counterfactualNanoUsd: number;
  counterfactualTurns: number;
}

/** Collects the receipt from every response that carries one. */
export class ReceiptCollector {
  private readonly observed: ObservedReceipt[] = [];
  private readonly onReceipt?: (receipt: ObservedReceipt) => void;
  private readonly limit: number;
  private readonly underlying: FetchLike;
  private seen = 0;
  private summed = 0;
  private summedCounterfactual = 0;
  private counterfactualTurns = 0;

  constructor(
    options: {
      /** Called as each receipt arrives. */
      onReceipt?: (receipt: ObservedReceipt) => void;
      /** How many receipts to keep in `all`. Default 1000; 0 keeps none. The total is exact regardless. */
      retain?: number;
      /** The fetch to wrap. Defaults to the global. */
      fetch?: FetchLike;
    } = {},
  ) {
    this.onReceipt = options.onReceipt;
    this.limit = options.retain ?? 1000;
    const base = options.fetch ?? (globalThis.fetch as unknown as FetchLike | undefined);
    if (base === undefined) {
      throw new Error("ReceiptCollector needs a fetch: this runtime has no global one, so pass `fetch`.");
    }
    this.underlying = base;
  }

  /** The wrapped fetch. A property, not a method, so `fetch: receipts.fetch` keeps `this`. */
  readonly fetch: FetchLike = async (url, init) => {
    const response = await this.underlying(url, init);
    const receipt = readReceipt(response.headers);
    if (receipt.costNanoUsd !== undefined || receipt.effectiveModel !== undefined) {
      this.record({ ...receipt, url: String(url), at: new Date() });
    }
    return response;
  };

  private record(receipt: ObservedReceipt): void {
    this.seen += 1;
    if (receipt.costNanoUsd !== undefined) this.summed += receipt.costNanoUsd;
    if (receipt.counterfactualNanoUsd !== undefined) {
      this.summedCounterfactual += receipt.counterfactualNanoUsd;
      this.counterfactualTurns += 1;
    }
    if (this.limit > 0) {
      this.observed.push(receipt);
      if (this.observed.length > this.limit) this.observed.shift();
    }
    // A throwing callback must not fail a turn the caller already paid for.
    try {
      this.onReceipt?.(receipt);
    } catch {
      /* the callback's problem */
    }
  }

  /** The retained receipts, oldest first. */
  get all(): readonly ObservedReceipt[] {
    return this.observed;
  }

  get last(): ObservedReceipt | undefined {
    return this.observed[this.observed.length - 1];
  }

  /** The running total over every turn seen, including any dropped from `all`. */
  get total(): ReceiptTotal {
    return {
      turns: this.seen,
      costNanoUsd: this.summed,
      costUsd: nanoUsdToUsdString(this.summed),
      counterfactualNanoUsd: this.summedCounterfactual,
      counterfactualTurns: this.counterfactualTurns,
    };
  }

  /** Forget the retained receipts and reset the total. */
  reset(): void {
    this.observed.length = 0;
    this.seen = 0;
    this.summed = 0;
    this.summedCounterfactual = 0;
    this.counterfactualTurns = 0;
  }
}

/**
 * A spend ceiling across many turns, enforced client-side. It refuses the
 * next request once the budget is spent; the worst case is budget plus one
 * turn, which a per-request `maxCostNanoUsd` bounds.
 */
export class SpendBudget {
  private readonly collector: ReceiptCollector;
  private readonly budgetNanoUsd: number;

  constructor(budgetNanoUsd: number, options: { fetch?: FetchLike } = {}) {
    if (!Number.isInteger(budgetNanoUsd) || budgetNanoUsd < 0) {
      throw new Error("a spend budget is a non-negative INTEGER nanodollar amount ($1 = 1e9).");
    }
    this.budgetNanoUsd = budgetNanoUsd;
    this.collector = new ReceiptCollector({ fetch: options.fetch, retain: 0 });
  }

  get spentNanoUsd(): number {
    return this.collector.total.costNanoUsd;
  }

  /** Never negative. */
  get remainingNanoUsd(): number {
    return Math.max(0, this.budgetNanoUsd - this.spentNanoUsd);
  }

  get exhausted(): boolean {
    return this.spentNanoUsd >= this.budgetNanoUsd;
  }

  /** The wrapped fetch: refuses once the budget is gone, then observes. */
  readonly fetch: FetchLike = async (url, init) => {
    if (this.exhausted) {
      throw new Error(
        `spend budget exhausted: ${nanoUsdToUsdString(this.spentNanoUsd)} USD spent against a ${nanoUsdToUsdString(this.budgetNanoUsd)} USD budget. This refusal is CLIENT-SIDE; the gateway was not called.`,
      );
    }
    return this.collector.fetch(url, init);
  };
}

File: python/conifer_sdk/receipts.py
"""Receipts for the client you already use. The OpenAI Python SDK takes an ``http_client`` and httpx exposes response hooks; this one reads the ``x-conifer-*`` headers on the way past and never touches the body."""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable, List, Mapping, Optional, Sequence

from .receipt import Receipt, nano_usd_to_usd_string, read_receipt


@dataclass
class ObservedReceipt:
    """One turn's receipt, plus where and when it came from."""

    receipt: Receipt
    #: The URL that was called.
    url: str
    #: When the response arrived.
    at: datetime

    @property
    def cost_nano_usd(self) -> Optional[int]:
        return self.receipt.cost_nano_usd

    @property
    def effective_model(self) -> Optional[str]:
        return self.receipt.effective_model


@dataclass
class ReceiptTotal:
    """A running total across every observed turn."""

    #: Number of turns that disclosed a receipt.
    turns: int
    #: Summed settled cost, in integer nanodollars.
    cost_nano_usd: int
    #: The same number as an exact decimal USD string.
    cost_usd: str
    #: Summed counterfactual over the turns that disclosed one; a subset, not a savings figure.
    counterfactual_nano_usd: int
    counterfactual_turns: int


class ReceiptCollector:
    """Collects the receipt from every response that carries one. Thread-safe."""

    def __init__(
        self,
        on_receipt: Optional[Callable[[ObservedReceipt], None]] = None,
        retain: int = 1000,
    ) -> None:
        #: Called as each receipt arrives.
        self._on_receipt = on_receipt
        #: How many receipts to keep in ``all``. 0 keeps none. The total is exact regardless.
        self._retain = retain
        self._observed: List[ObservedReceipt] = []
        self._lock = threading.Lock()
        self._seen = 0
        self._summed = 0
        self._summed_counterfactual = 0
        self._counterfactual_turns = 0

    # ------------------------------------------------------------ collection

    def observe(self, headers: Mapping[str, str], url: str = "") -> Optional[ObservedReceipt]:
        """Read a receipt off one response's headers. The generic entry point."""
        receipt = read_receipt(headers)
        if receipt.cost_nano_usd is None and receipt.effective_model is None:
            return None
        observed = ObservedReceipt(receipt=receipt, url=url, at=datetime.now(timezone.utc))
        self._record(observed)
        return observed

    def httpx_hook(self, response: Any) -> None:
        """An httpx ``response`` event hook."""
        self.observe(response.headers, str(getattr(response, "url", "")))

    def _record(self, observed: ObservedReceipt) -> None:
        with self._lock:
            self._seen += 1
            if observed.receipt.cost_nano_usd is not None:
                self._summed += observed.receipt.cost_nano_usd
            if observed.receipt.counterfactual_nano_usd is not None:
                self._summed_counterfactual += observed.receipt.counterfactual_nano_usd
                self._counterfactual_turns += 1
            if self._retain > 0:
                self._observed.append(observed)
                if len(self._observed) > self._retain:
                    self._observed.pop(0)
        # Outside the lock: a throwing callback must not fail a turn the caller already paid for.
        if self._on_receipt is not None:
            try:
                self._on_receipt(observed)
            except Exception:  # noqa: BLE001 - the callback's problem, not the request's
                pass

    # --------------------------------------------------------------- reading

    @property
    def all(self) -> Sequence[ObservedReceipt]:
        """The retained receipts, oldest first. A bounded tail; see ``retain``."""
        with self._lock:
            return list(self._observed)

    @property
    def last(self) -> Optional[ObservedReceipt]:
        """The most recent receipt, which is what a one-shot script wants."""
        with self._lock:
            return self._observed[-1] if self._observed else None

    @property
    def total(self) -> ReceiptTotal:
        """The running total over every turn seen, including any dropped from ``all``."""
        with self._lock:
            return ReceiptTotal(
                turns=self._seen,
                cost_nano_usd=self._summed,
                cost_usd=nano_usd_to_usd_string(self._summed),
                counterfactual_nano_usd=self._summed_counterfactual,
                counterfactual_turns=self._counterfactual_turns,
            )

    def reset(self) -> None:
        """Forget the retained receipts AND reset the total."""
        with self._lock:
            self._observed.clear()
            self._seen = 0
            self._summed = 0
            self._summed_counterfactual = 0
            self._counterfactual_turns = 0


class SpendBudget:
    """A hard spend ceiling across MANY turns, enforced client-side."""

    def __init__(self, budget_nano_usd: int) -> None:
        if not isinstance(budget_nano_usd, int) or isinstance(budget_nano_usd, bool):
            raise ValueError(
                "a spend budget is an INTEGER nanodollar amount ($1 = 1e9). Rounding a "
                "spend limit is the wrong direction half the time."
            )
        if budget_nano_usd < 0:
            raise ValueError("a spend budget cannot be negative")
        self.budget_nano_usd = budget_nano_usd
        self.collector = ReceiptCollector(retain=0)

    @property
    def spent_nano_usd(self) -> int:
        """Nanodollars spent so far, as observed on real receipts."""
        return self.collector.total.cost_nano_usd

    @property
    def remaining_nano_usd(self) -> int:
        """Nanodollars left. Never negative."""
        return max(0, self.budget_nano_usd - self.spent_nano_usd)

    @property
    def exhausted(self) -> bool:
        return self.spent_nano_usd >= self.budget_nano_usd

    def check(self) -> None:
        """Raise if the budget is gone. Call BEFORE issuing a request."""
        if self.exhausted:
            raise SpendBudgetExceeded(
                f"spend budget exhausted: {nano_usd_to_usd_string(self.spent_nano_usd)} USD "
                f"spent against a {nano_usd_to_usd_string(self.budget_nano_usd)} USD budget. "
                "This refusal is CLIENT-SIDE; the gateway was not called."
            )

    def httpx_hook(self, response: Any) -> None:
        """Observe a response's receipt against the budget."""
        self.collector.httpx_hook(response)


class SpendBudgetExceeded(RuntimeError):
    """The client-side multi-turn budget is spent. No request was made."""
